### ISSUE TITLE
Low performance `atand(y, x)`

### ISSUE THREAD

**Issue by kimikage** [JuliaLang/julia#41362]
The current implementation of `atand(y, x)` is to apply `rad2deg` to the result of `atan(y, x)`.
https://github.com/JuliaLang/julia/blob/ed4c44fbafbbace523ebdc9440bf1f542157048a/base/special/trig.jl#L1297-L1298
In this way, an error of at least 1 ULP is inevitable. However, the actual situation of `atand(::Float32, ::Float32)` is much worse, with a maximum error of more than 3 ULPs and a mean error of ~0.7 ULPs.
For example:
```julia
julia> y, x = 50.85106f0, -24.293373f0;

julia> f32 = atand(y, x)
115.53545f0

julia> bf = atand(big(y), big(x))
115.5354778291273278381248228732227454834884568423740673468650285334341643520974

julia> Float64((f32 - bf) / eps(Float32(bf)))
-3.1500193771143987
```
Of course, even if it has a low accuracy, it is beneficial if it is fast. Unfortunately, however, `atan` has a lot of conditional branches and is slow on modern CPUs.

Since in Colors.jl (ColorTypes.jl), hues are represented in degrees, not radians, the `atand` is an important function. So, I am proposing a more accurate and faster implementation. (cf. https://github.com/JuliaGraphics/Colors.jl/pull/495)

However, my proposed method still has a maximum error of about 2 ULPs. I think the Julia community places a lot of emphasis on accuracy when it comes to mathematical functions. I am not familiar with a high precision and fast implementation of `atan` or `atand`.

Does anyone have any ideas on improving `atand`?　Also, what degree of accuracy would you like to see in `atand`?

**Comment by oscardssmith** [comment]
Fixing this should be relatively easy. Basically, all that needs to happen is copy the code for atan, and change the constants so it calculates atand. Accuracy of 1.5 ulps and improved speed should be relatively easy to achieve here. The only complicated part is keeping the code clean.

**Comment by kimikage** [comment]
Fortunately, π is close to `180/64`. Perhaps the algorithm can be shared with `atan`.
However, with the current implementation of `atan`, I don't think it will be easy to speed it up.

**Comment by oscardssmith** [comment]
Yeah. I think we need a rewrite of both the 1 and 2 arg atan. They both could be a bunch faster

**Comment by oscardssmith** [comment]
For 1 arg atan, I believe we want to separate into x<sqrt(2)-1, x>sqrt(2)+1, and the middle. Big x can be handled as pi/2-atan(1/x) and medium x can be handled as atan(1)+atan((x-1)/(1+x)).  Atan with 2 arguments can be sped up more by not calling the 1 arg version, and optimizing the special case handling. The degree based version can then use the same logic with different constants to make it have no overhead (and possibly better accuracy)

**Comment by kimikage** [comment]
FWIW, in https://github.com/JuliaGraphics/Colors.jl/pull/495, the kernel part is used in the range [0°, 26.6°] due to the balance between maximum error and speed. As suggested by @oscardssmith, it is easier to improve the mean accuracy if the range is [-22.5°, 22.5°] rather than [-26.6°, 26.6°].

This is not an urgent issue for me. (In any case, ColorTypes.jl and Colors.jl need to support the Julia LTS version.) Also, I don't have the sufficient time to submit a PR and do some tweaking. :tired_face:

Of course, anyone can tackle this issue.

**Comment by oscardssmith** [comment]
See https://github.com/JuliaLang/julia/pull/41371 for the start on this.
